File: packages/jev-router/facts.mjs
/**
 * Deterministic facts + available() gates. Hard eligibility before Jev route.
 * No Stanley binary. review is a routing stub this spike (not implemented).
 */

import { execFileSync } from "node:child_process";
import { existsSync, readFileSync } from "node:fs";
import { CHECK_POLICY } from "./policy.mjs";

export const WORKFLOWS = {
  check: {
    id: "check",
    implemented: true,
    available: (facts) => facts.diffPresent === true,
    unavailableReason: "check requires a git diff",
  },
  review: {
    id: "review",
    implemented: false,
    available: (facts) => facts.diffPresent === true,
    unavailableReason: "review requires a git diff (stub; not implemented this spike)",
  },
};

export function diffPresent(diffText) {
  return String(diffText || "").trim().length > 0;
}

export function availableCapabilities(facts) {
  const capabilities = {};
  const unavailable = {};
  for (const workflow of Object.values(WORKFLOWS)) {
    const ok = workflow.available(facts);
    capabilities[workflow.id] = ok;
    if (!ok) unavailable[workflow.id] = workflow.unavailableReason;
  }
  return { capabilities, unavailable };
}

export function availableIds(facts) {
  return Object.entries(availableCapabilities(facts).capabilities)
    .filter(([, ok]) => ok)
    .map(([id]) => id);
}

function runGit(repo, args, exec = execFileSync) {
  return exec("git", args, {
    cwd: repo,
    encoding: "utf8",
    stdio: ["ignore", "pipe", "pipe"],
    timeout: 15_000,
  });
}

/**
 * Read a unified diff from a file, git, or injected text.
 * Secrets stay out of argv: TYPESAFE_API_KEY is never read here.
 */
export function gatherDiff({
  repo = process.cwd(),
  base = "",
  diffFile = "",
  diffText = null,
  exec = execFileSync,
} = {}) {
  if (diffText != null) {
    const text = String(diffText);
    return { text, present: diffPresent(text), source: "injected", error: null };
  }
  if (diffFile) {
    if (!existsSync(diffFile)) {
      return { text: "", present: false, source: "file", error: `diff file not found: ${diffFile}` };
    }
    const text = readFileSync(diffFile, "utf8");
    return { text, present: diffPresent(text), source: "file", error: null };
  }

  try {
    const args = ["diff", "--no-color", "--no-ext-diff"];
    if (base) args.push(base);
    const vsBaseOrWorktree = runGit(repo, args, exec);
    const staged = base ? "" : runGit(repo, ["diff", "--cached", "--no-color", "--no-ext-diff"], exec);
    const text = [vsBaseOrWorktree, staged].filter((part) => part && part.trim()).join("\n");
    return { text, present: diffPresent(text), source: base ? `git-diff:${base}` : "git-worktree", error: null };
  } catch (err) {
    const message = err && err.message ? err.message : String(err);
    return { text: "", present: false, source: "git", error: message };
  }
}

/** Clip write evidence to the check policy hunk cap. Default is 4000. */
export function clipToMaxHunkChars(text, max = CHECK_POLICY.maxHunkChars) {
  const n = Number(max);
  const cap = Number.isFinite(n) && n >= 0 ? Math.trunc(n) : CHECK_POLICY.maxHunkChars;
  return String(text ?? "").slice(0, cap);
}

function writePath(input) {
  return String(input.path || input.file_path || input.filePath || "").replace(/[\r\n]/g, "");
}

function hasOwn(obj, key) {
  return Object.prototype.hasOwnProperty.call(obj, key);
}

function diffHeader(path, created) {
  const name = path || "unknown";
  return [`diff --git a/${name} b/${name}`, created ? "--- /dev/null" : `--- a/${name}`, `+++ b/${name}`];
}

function prefixedLines(prefix, value) {
  return String(value ?? "")
    .split(/\r?\n/)
    .map((line) => `${prefix}${line}`);
}

/**
 * Unified diff of a write tool body. Empty when the input is only a path.
 * This is not the git worktree. `--check` still uses gatherDiff.
 */
export function proposedWriteDiff(input) {
  if (typeof input === "string") return { path: input, text: "" };
  if (!input || typeof input !== "object" || Array.isArray(input)) return { path: "", text: "" };
  const path = writePath(input);
  if (Array.isArray(input.edits)) {
    if (input.edits.length === 0) return { path, text: "" };
    const parts = diffHeader(path, false);
    for (const edit of input.edits) {
      const row = edit && typeof edit === "object" ? edit : {};
      parts.push("@@", ...prefixedLines("-", row.old_string ?? ""), ...prefixedLines("+", row.new_string ?? ""));
    }
    return { path, text: parts.join("\n") };
  }
  if (hasOwn(input, "old_string") || hasOwn(input, "new_string")) {
    const parts = diffHeader(path, false);
    parts.push("@@", ...prefixedLines("-", input.old_string ?? ""), ...prefixedLines("+", input.new_string ?? ""));
    return { path, text: parts.join("\n") };
  }
  if (hasOwn(input, "contents")) {
    const parts = diffHeader(path, true);
    parts.push("@@", ...prefixedLines("+", input.contents ?? ""));
    return { path, text: parts.join("\n") };
  }
  return { path, text: "" };
}

/**
 * Hook wire for a write. A body becomes JSON `{path, proposed}` clipped to
 * maxHunkChars. A path with no body stays the path string.
 */
export function writeToolWire(input, max = CHECK_POLICY.maxHunkChars) {
  const { path, text } = proposedWriteDiff(input);
  if (!text) return path;
  return JSON.stringify({ path, proposed: clipToMaxHunkChars(text, max) });
}

/** Inverse of writeToolWire. Non-JSON input is a path and an empty diff. */
export function parseWriteToolWire(raw, max = CHECK_POLICY.maxHunkChars) {
  const text = String(raw ?? "");
  if (text.startsWith("{")) {
    try {
      const parsed = JSON.parse(text);
      if (parsed && typeof parsed === "object" && !Array.isArray(parsed) && typeof parsed.proposed === "string") {
        return {
          path: String(parsed.path || "").replace(/[\r\n]/g, ""),
          diffText: clipToMaxHunkChars(parsed.proposed, max),
        };
      }
    } catch {
      /* A path, not an envelope. */
    }
  }
  return { path: text, diffText: "" };
}

export function gatherFacts(opts = {}) {
  const diff = gatherDiff(opts);
  const { capabilities, unavailable } = availableCapabilities({ diffPresent: diff.present });
  return {
    diffPresent: diff.present,
    diffSource: diff.source,
    diffError: diff.error,
    capabilities,
    unavailable,
    available: Object.entries(capabilities)
      .filter(([, ok]) => ok)
      .map(([id]) => id),
  };
}

File: crates/bezel-bridle/src/facts.rs
//! Deterministic facts. Proposed edits are clipped. `--check` still reads git.
//! No client is constructed here.

use crate::policy::MAX_HUNK_CHARS;
use std::path::{Path, PathBuf};
use std::process::Command;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DiffFacts {
    pub text: String,
    pub present: bool,
    pub source: String,
    pub error: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Edit {
    pub old_string: String,
    pub new_string: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct WriteBody {
    pub path: String,
    pub file_path: Option<String>,
    pub old_string: Option<String>,
    pub new_string: Option<String>,
    pub contents: Option<String>,
    pub edits: Option<Vec<Edit>>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Gathered {
    pub proposed: String,
    pub diff: DiffFacts,
    pub available: Vec<String>,
}

#[derive(Debug, Clone)]
pub struct GatherOpts {
    pub repo: PathBuf,
    pub base: String,
    pub diff_file: Option<PathBuf>,
    pub diff_text: Option<String>,
    pub write: Option<WriteBody>,
}

impl Default for GatherOpts {
    fn default() -> Self {
        Self {
            repo: PathBuf::from("."),
            base: String::new(),
            diff_file: None,
            diff_text: None,
            write: None,
        }
    }
}

pub fn clip_to_max_hunk_chars(text: &str, max: Option<usize>) -> String {
    let cap = max.unwrap_or(MAX_HUNK_CHARS);
    text.chars().take(cap).collect()
}

pub fn proposed_write_diff(input: &WriteBody) -> (String, String) {
    let path = write_path(input);
    if let Some(edits) = &input.edits {
        if edits.is_empty() {
            return (path, String::new());
        }
        let mut parts = diff_header(&path, false);
        for edit in edits {
            parts.push("@@".to_string());
            parts.extend(prefixed_lines("-", &edit.old_string));
            parts.extend(prefixed_lines("+", &edit.new_string));
        }
        return (path, parts.join("\n"));
    }
    if input.old_string.is_some() || input.new_string.is_some() {
        let mut parts = diff_header(&path, false);
        parts.push("@@".to_string());
        parts.extend(prefixed_lines("-", input.old_string.as_deref().unwrap_or("")));
        parts.extend(prefixed_lines("+", input.new_string.as_deref().unwrap_or("")));
        return (path, parts.join("\n"));
    }
    if let Some(contents) = &input.contents {
        let mut parts = diff_header(&path, true);
        parts.push("@@".to_string());
        parts.extend(prefixed_lines("+", contents));
        return (path, parts.join("\n"));
    }
    (path, String::new())
}

pub fn write_tool_wire(input: &WriteBody) -> String {
    let (path, text) = proposed_write_diff(input);
    if text.is_empty() {
        return path;
    }
    let proposed = clip_to_max_hunk_chars(&text, None);
    format!(
        "{{\"path\":{},\"proposed\":{}}}",
        json_string(&path),
        json_string(&proposed)
    )
}

pub fn gather_diff(opts: &GatherOpts) -> DiffFacts {
    if let Some(text) = &opts.diff_text {
        return DiffFacts {
            present: diff_present(text),
            text: text.clone(),
            source: "injected".to_string(),
            error: None,
        };
    }
    if let Some(path) = &opts.diff_file {
        if !path.exists() {
            return DiffFacts {
                text: String::new(),
                present: false,
                source: "file".to_string(),
                error: Some(format!("diff file not found: {}", path.display())),
            };
        }
        let text = std::fs::read_to_string(path).unwrap_or_default();
        return DiffFacts {
            present: diff_present(&text),
            text,
            source: "file".to_string(),
            error: None,
        };
    }
    match git_worktree(&opts.repo, &opts.base) {
        Ok(text) => DiffFacts {
            present: diff_present(&text),
            source: if opts.base.is_empty() {
                "git-worktree".to_string()
            } else {
                format!("git-diff:{}", opts.base)
            },
            text,
            error: None,
        },
        Err(error) => DiffFacts {
            text: String::new(),
            present: false,
            source: "git".to_string(),
            error: Some(error),
        },
    }
}

/// Shipped gather: clipped proposed edit plus the git or file diff.
pub fn gather(opts: &GatherOpts) -> Gathered {
    let diff = gather_diff(opts);
    let proposed = opts.write.as_ref().map(write_tool_wire).unwrap_or_default();
    let available = if diff.present {
        vec!["check".to_string(), "review".to_string()]
    } else {
        Vec::new()
    };
    Gathered {
        proposed,
        diff,
        available,
    }
}

fn diff_present(text: &str) -> bool {
    !text.trim().is_empty()
}

fn write_path(input: &WriteBody) -> String {
    let raw = if !input.path.is_empty() {
        input.path.as_str()
    } else {
        input.file_path.as_deref().unwrap_or("")
    };
    raw.replace(['\r', '\n'], "")
}

fn diff_header(path: &str, created: bool) -> Vec<String> {
    let name = if path.is_empty() { "unknown" } else { path };
    vec![
        format!("diff --git a/{name} b/{name}"),
        if created {
            "--- /dev/null".to_string()
        } else {
            format!("--- a/{name}")
        },
        format!("+++ b/{name}"),
    ]
}

fn prefixed_lines(prefix: &str, value: &str) -> Vec<String> {
    value.split('\n').map(|line| {
        let line = line.strip_suffix('\r').unwrap_or(line);
        format!("{prefix}{line}")
    }).collect()
}

fn json_string(value: &str) -> String {
    let mut out = String::from("\"");
    for ch in value.chars() {
        match ch {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\n' => out.push_str("\\n"),
            '\r' => out.push_str("\\r"),
            '\t' => out.push_str("\\t"),
            c if (c as u32) < 0x20 => out.push_str(&format!("\\u{:04x}", c as u32)),
            c => out.push(c),
        }
    }
    out.push('"');
    out
}

fn git_worktree(repo: &Path, base: &str) -> Result<String, String> {
    let mut args = vec!["diff", "--no-color", "--no-ext-diff"];
    if !base.is_empty() {
        args.push(base);
    }
    let work = run_git(repo, &args)?;
    let staged = if base.is_empty() {
        run_git(repo, &["diff", "--cached", "--no-color", "--no-ext-diff"])?
    } else {
        String::new()
    };
    Ok([work, staged]
        .into_iter()
        .filter(|part| !part.trim().is_empty())
        .collect::<Vec<_>>()
        .join("\n"))
}

fn run_git(repo: &Path, args: &[&str]) -> Result<String, String> {
    let run = Command::new("git")
        .args(args)
        .current_dir(repo)
        .env("GIT_CONFIG_GLOBAL", "/dev/null")
        .env("GIT_CONFIG_SYSTEM", "/dev/null")
        .output()
        .map_err(|err| format!("git {} failed to spawn: {err}", args.join(" ")))?;
    if !run.status.success() {
        return Err(format!(
            "git {} failed: {}",
            args.join(" "),
            String::from_utf8_lossy(&run.stderr).trim()
        ));
    }
    String::from_utf8(run.stdout).map_err(|err| err.to_string())
}
